### ISSUE TITLE
Cross-network non-determinism on pool-allocated tensors persists after #315 zero-padding fix (was: #311 DBM Clone drift)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#318]
## Summary

Issue #311 was closed by [#315](https://github.com/ooples/AiDotNet.Tensors/pull/315) ("zero pooled-tensor padding so SIMD overhang reads can't leak garbage"), but the empirical bug it filed against still reproduces unchanged on AiDotNet.Tensors 0.75.3. Filing as a separate issue per maintainer request rather than reopening since the original was closed against a specific PR that did land.

## What persists post-#315

A `Tensor<T>` constructed via training-time engine ops (e.g. `Engine.TensorSubtract` consuming and producing pool-allocated tensors during `UpdateParameters`) and a `Tensor<T>` populated via `SetFlat` writes into a fresh allocation with byte-identical logical data produce **different** output when both are subsequently passed through `Engine.TensorMatMul` / `Engine.FusedLinear` / `Engine.TensorTranspose`.

This breaks the standard PyTorch/TensorFlow `model.state_dict()` round-trip contract: same parameters → same forward output.

## Reproduction

Tested on consumer repo `ooples/AiDotNet` PR #1224 (PR #1274 working branch). Standalone harness at `tools/DbmCloneDebug/Program.cs`:

```csharp
var network = new DeepBoltzmannMachine<double>();    // Layers: RBM + RBM + Dense, 379,756 params
// 10 train iterations on a fixed (input, target) pair using TrainWithTape →
// _weights / _visibleBiases / _hiddenBiases on each layer get rebound to engine-op
// outputs (Engine.TensorSubtract products), so they live in the SIMD-padded pool.

var trainedOut = network.Predict(probe);
var trainedParams = network.GetParameters();         // walks Layers, returns flat Vector<T>

var cloned = (DeepBoltzmannMachine<double>)network.Clone();
// Clone goes via DeepCopy → SerializeInternalUnchecked → DeserializeInternalUnchecked,
// which calls layer.SetParameters(Vector<T>) on the cloned layers. RBMLayer.SetParameters
// uses _weights.SetFlat(i, value) — writes into the fresh-allocated buffer at construction.

var clonedParams = cloned.GetParameters();
var clonedOut = cloned.Predict(probe);
```

### Result on 0.75.3

```
Tensors 0.75.3 diagnostic:
  Param count: 379756, max param diff: 0.000E+000   ← byte-identical params
  Out diff: max abs = 3.508E-002                    ← 3.5% output drift
    out[0]: trained=0.10870797  cloned=0.10517597  diff=3.532E-003
    out[1]: trained=0.07304232  cloned=0.06283829  diff=1.020E-002
    out[2]: trained=0.91160650  cloned=0.91561788  diff=4.011E-003
    out[3]: trained=0.27038165  cloned=0.26825793  diff=2.124E-003
    out[4]: trained=0.12109474  cloned=0.09941344  diff=2.168E-002
```

### Drift across versions (same harness, same hardware)

| Tensors version | Output diff (max abs) |
|---|---:|
| 0.72.0 | 3.6 % |
| 0.75.2 | 3.6 % |
| **0.75.3 (#315 zero-pad fix landed)** | **3.5 %** |

Magnitude is essentially constant — #315 didn't move the needle.

## Self-determinism is fine

```
Self-determinism: original max-diff=0.000E+000, cloned max-diff=0.000E+000
```

Two consecutive `Predict(probe)` calls on the **same** network produce bit-identical output. So the underlying ops *are* deterministic functions of their tensor operands. The difference is purely a function of how `_weights` was constructed.

## Hypothesis (still load-bearing post-#315)

Either:

1. **Padding bytes still vary by allocation site.** #315 zeroes padding at allocation time, but downstream engine ops may write through-or-past the logical extent (SIMD prologue/epilogue), leaving non-zero padding bytes that survive into the next consumer. A tensor hot off `Engine.TensorSubtract` could end up with non-zero padding, while a freshly-allocated tensor that's only seen `SetFlat(i, value)` writes (which only touch logical indices) would have zero padding. Subsequent `Engine.TensorMatMul` or `Engine.TensorTranspose` reading SIMD lanes that include padding sees different data → different reduction order → different output.

2. **SIMD reduction order depends on alignment, not just contents.** If `_weights` from training is allocated at a different-aligned offset within the pool than `_weights` from `new Tensor<T>(shape)` at construction, the SIMD vector loop in `SimdGemm.DgemmAvx2` (or wherever) may use different lane orderings even with byte-identical contents, and float64 reduction is non-associative.

3. **Some non-content tensor metadata is differing.** Stride, view-info, scratch-buffer pointer, etc. — anything carried alongside `Memory<T>` that downstream kernels consult to decide their codepath.

The empirical signature (byte-identical contents, deterministic same-network `Predict`, but cross-network `Predict` divergence) narrows it to one of these.

## Suggested investigation

- Run a heap dump on the two `_weights` tensors and diff them at the byte level (logical + padding region). Are padding bytes identical post-#315?
- If yes: investigate the kernel path. `SimdGemm.DgemmAvx2` etc. — does it consult any tensor metadata other than `Length` / `Memory`?
- If no: #315's zero-on-allocate is being overwritten somewhere. Track who's writing into the padding region.

## What would close this issue

A user-callable `Engine.Canonicalize(Tensor<T>)` op (analog of PyTorch `.contiguous()`) that maintainers guarantee makes the tensor's full backing state byte-equal regardless of construction path. Then consumers can call it after `SetFlat`-style population in `SetParameters` and the round-trip contract holds.

Or: kernel-side fix that makes `Engine.TensorMatMul` / `FusedLinear` / `TensorTranspose` provably layout-invariant.

## Why it matters for AiDotNet

`DeepBoltzmannMachineTests.Clone_AfterTraining_ShouldPreserveLearnedWeights` and `BlendedDiffusionModelTests.Clone_ShouldProduceIdenticalOutput` (and likely many other long-forward-chain models) fail on this. The `||Δ||` magnitudes are 1.65E-1 (DBM, fails tolerance 6.77E-5) and similar across other models with 3+ matmul layers in the forward path.

This was filed in #311 with [comment 4410978556](https://github.com/ooples/AiDotNet.Tensors/issues/311#issuecomment-4410978556) showing the persistence; raising as a fresh issue for visibility.

**Comment by ooples** [comment]
## Resolution: false trail — not a Tensors bug

After deeper investigation in the consumer repo, this issue (and the original #311) is **not a Tensors-side bug**. The cross-network non-determinism is a stale-cache invalidation gap in `AiDotNet.NeuralNetworks.Layers.RBMLayer`.

### Diagnostic chain

I wrote three Tensors-side kernel-level repros:

1. **`Issue318KernelLevelReproTests.TensorMatMul_FreshVsEngineOpLhs_ProducesIdenticalOutput`** — same lhs values via two paths (fresh + SetFlat / `TensorSubtract(zero, neg)`), same rhs, two `TensorMatMul` calls. Result: **0.000E+000 diff**.

2. **`Issue318TransposeReproTests.TensorTranspose_FreshVsEngineOp_ProducesIdenticalOutput`** — same input via two paths, two `TensorTranspose` calls. Result: **0.000E+000 diff**.

3. **`Issue318FullChainReproTests.ThreeLayerChain_FreshVsEngineOpWeights_ProducesIdenticalOutput`** — full DBM-faithful 3-layer chain (128 → 500 → 500 → 128) with two parallel weight construction paths. Result: **0.000E+000 diff**.

All three pass on Tensors 0.75.3. The kernels are deterministic and layout-invariant.

### What was actually broken

The consumer-side drift was caused by this in `RBMLayer.Forward`:

```csharp
// Use FusedLinear for tape-tracked forward: h = sigmoid(v @ W^T + b_h)
_weightsTCache ??= Engine.TensorTranspose(_weights);
Tensor<T> hiddenProbs = Engine.FusedLinear(visible2D, _weightsTCache, ...);
```

`_weightsTCache` was invalidated only via `RBMLayer.UpdateParameters` and the contrastive-divergence path. The tape-based training path (`TrainWithTape` → `optimizer.Step` on `ParameterBuffer`) updates the underlying buffer in place **without** calling the layer's `UpdateParameters`. So the cache stayed at a stale (one-optimizer-step-old) transpose for the entire training run.

Reflection probe of the consumer DBM after train + first predict:
```
A (after 10 iter Train): _weights[0]=-0.054754632996, _weightsTCache=-0.053755352930  ← NON-NULL
B (after first Predict): _weights[0]=-0.054754632996, _weightsTCache=-0.053755352930  ← STALE
```

The cached `_weightsTCache[0]` was off by exactly 0.001 from the live `_weights[0]` value — that's the Adam learning rate. When the `??=` saw the cache non-null on subsequent forwards, it skipped recomputing. Original network's predict consumed the stale cache; cloned network's fresh `_weightsTCache=null` forced a recompute → fresh transpose → different forward output → ~3.5% drift through the 3-layer chain.

### Fix (consumer-side, ooples/AiDotNet)

Removed the cache field. `RBMLayer.Forward` now recomputes `Engine.TensorTranspose(_weights)` every call. Cost: one O(visibleUnits × hiddenUnits) transpose per forward, negligible relative to the FusedLinear that follows it. Matches how every other tape-trained layer handles weight reads.

Verified:
- `tools/DbmCloneDebug` standalone diagnostic: drift 3.5% → **0.000E+000**.
- `DeepBoltzmannMachineTests`: 20/21 → **21/21** passing on Tensors 0.75.3.

Commit: [ooples/AiDotNet@5ea3cecd3](https://github.com/ooples/AiDotNet/commit/5ea3cecd3).

### Implication for #318/#311 + PR #320

This issue can be **closed as not-a-bug** in Tensors. PR #320's `Canonicalize()` API isn't needed — the kernels are already layout-invariant. The fact that I couldn't reproduce the drift in pure-Tensors test scope was correct — the drift required RBMLayer's caching pattern AND the tape-trained ParameterBuffer update path, neither of which the Tensors test suite has access to.

Apologies for the wrong-trail diagnosis in the original report; the symptom (byte-identical params, divergent forward output) genuinely looked like a kernel determinism issue from the outside, but the actual root cause was a missing cache-invalidation hook in the consumer's layer.

**Comment by ooples** [comment]
Closing as not-a-bug per [comment above](#issuecomment-) — root cause was a stale-cache invalidation gap in consumer-side RBMLayer (ooples/AiDotNet@5ea3cecd3), not a Tensors-side issue. Kernel-level repros verify TensorMatMul / TensorTranspose / FusedLinear are layout-invariant on 0.75.3.
